odbc/src/api/data.rs

```diff
@@ -0,0 +1,647 @@
+/// should be stored in [`text_length_ptr`].
+///
+/// # Safety
+/// This writes to multiple raw C-pointers
+///
+pub unsafe fn set_output_string(
+    message: &str,
+    output_ptr: *mut Char,
+    buffer_len: usize,
+    text_length_ptr: *mut SmallInt,
+) -> SqlReturn {
+    if output_ptr.is_null() {
+        if !text_length_ptr.is_null() {
+            // If the output_ptr is NULL, we should still return the length of the message.
+            *text_length_ptr = message.len() as i16;
+        }
+        return SqlReturn::SUCCESS_WITH_INFO;
+    }
+    // Check if the entire message plus a null terminator can fit in the buffer;
+    // we should truncate the message if it's too long.
+    // Note, we also assume this is valid ascii
+    let mut message_u8 = message.bytes().collect::<Vec<u8>>();
+    let message_len = message_u8.len();
+    let num_chars = min(message_len + 1, buffer_len);
+    // It is possible that no buffer space has been allocated.
+    if num_chars == 0 {
+        return SqlReturn::SUCCESS_WITH_INFO;
+    }
+    message_u8.resize(num_chars - 1, 0);
+    message_u8.push(0u8);
+    copy_nonoverlapping(message_u8.as_ptr(), output_ptr, num_chars);
+    // Store the number of characters in the message string, excluding the
+    // null terminator, in text_length_ptr
+    if !text_length_ptr.is_null() {
+        *text_length_ptr = (num_chars - 1) as SmallInt;
+    }
+    if num_chars < message_len {
+        SqlReturn::SUCCESS_WITH_INFO
+    } else {
+        SqlReturn::SUCCESS
+    }
+}
+
+///
+/// get_diag_rec copies the given ODBC error's diagnostic information
+/// into the provided pointers.
+///
+/// # Safety
+/// This writes to multiple raw C-pointers
+///
+pub unsafe fn get_diag_rec(
+    error: &ODBCError,
+    state: *mut WChar,
+    message_text: *mut WChar,
+    buffer_length: SmallInt,
+    text_length_ptr: *mut SmallInt,
+    native_error_ptr: *mut Integer,
+) -> SqlReturn {
+    if !native_error_ptr.is_null() {
+        *native_error_ptr = error.get_native_err_code();
+    }
+    set_sql_state(error.get_sql_state(), state);
+    let message = format!("{}", error);
+    set_output_wstring(
+        &message,
+        message_text,
+        buffer_length as usize,
+        text_length_ptr,
+    )
+}
+
+///
+/// unsupported_function is a helper function for correctly setting the state for
+/// unsupported functions.
+///
+pub fn unsupported_function(handle: &mut MongoHandle, name: &'static str) -> SqlReturn {
+    handle.clear_diagnostics();
+    handle.add_diag_info(ODBCError::Unimplemented(name));
+    SqlReturn::ERROR
+}
+
+///
+/// set_str_length writes the given length to [`string_length_ptr`].
+///
+/// # Safety
+/// This writes to a raw C-pointers
+///
+pub unsafe fn set_str_length(string_length_ptr: *mut Integer, length: Integer) {
+    if !string_length_ptr.is_null() {
+        *string_length_ptr = length
+    }
+}
+
+#[cfg(test)]
+mod unit {
```

Comment:
They don't operate on actual bson values. These tests will be moved once Decimal128 is actually supported.

Reply:
Actually most of these tests will go away entirely in favor of just one test for each target type with a Decimal128 value, since we'll no longer need coverage for multiple types of Decimal128 (since that will be in the rust driver tests).

Reply:
oh ok, that sounds good 👍